projects/npm-tools/packages/npm-scripts/src/scripts/scanExposes.js

```diff
@@ -0,0 +1,109 @@
+/**
+ * SPDX-FileCopyrightText: © 2019 Liferay, Inc. <https://liferay.com>
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+/* eslint-disable no-console */
+
+const fs = require('fs');
+const path = require('path');
+
+const getMergedConfig = require('../utils/getMergedConfig');
+
+/**
+ * Main script that scans JSP files for JS modules and writes a proposed
+ * `exposes` section for the `npmscripts.config.js` file.
+ */
+module.exports = async function (...args) {
+	const {input} = getMergedConfig('npmscripts', 'build');
+
+	const update = args.includes('-u');
+	const verbose = args.includes('-v');
+
+	const exposes = scanJSPs(input, verbose);
+
+	if (verbose) {
+		console.log('');
+	}
+
+	console.log(exposes);
+
+	if (update) {
+		try {
+			let config = fs.readFileSync('npmscripts.config.js', 'utf8');
+
+			config +=
+				'module.exports.federation = module.exports.federation || {};\n';
+			config += 'module.exports.federation.exposes = [\n';
+
+			exposes.forEach((expose) => {
+				config += `  '${expose}',\n`;
+			});
+
+			config += '];';
```

Comment:
Hm. This seems like it could mangle an existing config file pretty badly (or least, in a way that no human would ever write it). For example, if I already have a `federation` config, this will overwrite it, leaving the other one behind. We have a few examples of AST-based autofixes in this repo that you could use as a model if you wanted to do it "The Right Way™".

Reply:
This file was intended to be removed. I'll push another commit to do it.